### ISSUE TITLE
[dynamo] Mutations of a closure-captured list/dict are silently dropped when they happen in a method of a locally defined class instantiated inside the compiled function (`fullgraph=True`, no graph break)

### ISSUE THREAD

**Issue by cuiliaomei-beep** [pytorch/pytorch#197896]
### 🐛 Describe the bug

**Summary.** When a compiled function instantiates a locally defined class and calls a method that mutates a list or dict captured from the enclosing scope, eager mutates it on every call and `torch.compile` never does: `list.append`, `dict[k] = v` and `dict[k] += 1` leave the container unchanged, while the tensor result is correct (torch 2.14.0 and nightly; `backend="eager"` with `fullgraph=True` is enough). A helper object that counts calls or records shapes leaves an empty log after compiling, with no graph break and no error, and a context manager written this way loses its state. Because the mutation is replayed correctly when the object is created outside the compiled function or a nested function is used rather than a method, the loss is specific to method closures of objects created while tracing.

```python
import torch

def make(kind):
    state = {"n": 0} if kind.startswith("dict") else []
    class Helper:
        def touch(self):
            if kind == "dict +=":
                state["n"] += 1
            elif kind == "dict setitem":
                state["k"] = 1
            elif kind == "list append":
                state.append(1)
            elif kind == "list +=":
                state.extend([1])
    def f(x):
        Helper().touch()
        return x + 1
    return f, state

for kind in ("dict +=", "dict setitem", "list append", "list +="):
    res = []
    for comp in (False, True):
        torch._dynamo.reset()
        f, state = make(kind)
        g = torch.compile(f, backend="eager", fullgraph=True) if comp else f
        g(torch.ones(1)); g(torch.ones(1))
        res.append(str(state))
    print(f"{kind:13s} eager {res[0]:18s} compiled {res[1]}")
```

```
dict +=       eager {'n': 2}           compiled {'n': 0}
dict setitem  eager {'n': 0, 'k': 1}   compiled {'n': 0}
list append   eager [1, 1]             compiled []
list +=       eager [1, 1]             compiled []
```

Forms that behave correctly with the same `Helper` and the same `state` (from `closure_mutation_minimal.py`):

| call inside the compiled function | compiled result |
|---|---|
| `Helper().touch()` (object created inside `f`) | **mutation lost** |
| `outer.touch()` with `outer = Helper()` created outside `f` | correct |
| a nested function `helper_fn()` that performs the same mutation | correct |
| class and state at module level | correct |

The same defect makes a locally defined context manager lose its side effects:

```python
def run():
    log = []
    class CM:
        def __enter__(self): log.append("enter"); return self
        def __exit__(self, *a): log.append("exit"); return False
    def f(x):
        with CM():
            y = x + 1
        return y * 2
    torch.compile(f, backend="eager")(torch.ones(2))
    return log          # eager: ['enter', 'exit']   compiled: []
```

Expected: the mutation is replayed after the graph runs, as it is for the other three forms; or, if Dynamo cannot track the closure cell of a method of an object without a source, a graph break (an error under `fullgraph=True`) instead of silently dropping the effect.

A likely place to look: the receiver is a `UserDefinedObjectVariable` created during tracing and therefore has no source; the closure cell reached through its method appears to be tracked as a fresh local object rather than as the pre-existing cell of the enclosing frame, so its mutation is never registered with `SideEffects` for replay.

### Error logs

No error and no graph break; the side effect is missing as shown above.

### Versions

```
PyTorch version: 2.14.0+cpu
OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
Python version: 3.14.7 (64-bit runtime)
Is CUDA available: False
CPU: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
```

Same output on nightly (2.15.0.dev20260911 on Windows, 2.15.0.dev20260920 on Linux) and on Linux with 2.14.0 (Python 3.12), so it does not depend on the Python version or the platform.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98

**Comment by syerva-g** [comment]
Hi! I'd like to pick up this issue — I have a fix ready and will put up a PR shortly.

### Root Cause & Approach
When a class (`Helper()`) or context manager (`with CM():`) is instantiated inside a `@torch.compile` region, the resulting `UserDefinedObjectVariable` is sourceless (`self.source is None`), while its class provenance is tracked on `self.cls_source`.

Currently, when resolving and calling methods on such instances, the unbound method's `source_fn` (`DictGetItemSource(TypeDictSource(cls_source), method_name)`) is lost in three places:
1. `UserDefinedObjectVariable.resolve_type_attr` checks `can_use_mro_source = self.cls_source is not None and self.source is not None` (and `call_method` / `resolve_data_descriptor` similarly gate `get_source_by_walking_mro` on `self.source` instead of `self.cls_source`), leaving `source_fn=None` on the created `UserMethodVariable`.
2. `UserFunctionVariable.get_source()` checks `if source and isinstance(self, variables.UserMethodVariable):`, so when `self.source is None` (newly constructed instance) and `self.source_fn is not None`, `get_source()` returns `None` instead of `self.source_fn`.
3. `GenericContextWrappingVariable.enter` and `exit` do not pass `source_fn = self.get_source_by_walking_mro(tx, ...)` when `self.cls_source is not None`.

Because `get_source()` returns `None`, `UserFunctionVariable.bind_args` builds the method's closure cells without a `Source`, causing `SideEffects` to track closure-captured lists/dicts as `MutationNew` and silently drop in-place mutations instead of emitting ` side_effects` bytecode.

**Fix:**
- Allow `get_source_by_walking_mro` whenever `self.cls_source is not None` in `UserDefinedObjectVariable` (`resolve_type_attr`, `resolve_data_descriptor`, `call_method`) and `GenericContextWrappingVariable` (`enter`, `exit`).
- Return `self.source_fn` in `UserFunctionVariable.get_source()` whenever `isinstance(self, variables.UserMethodVariable) and self.source_fn is not None`.
- Register closure cells in `side_effects.track_cell_existing` when `allow_sourced_cells=True` in `NestedUserFunctionVariable.get_function` (`__build_class__` path).
